tiledb/sm/filter/filter_pipeline.cc

```diff
@@ -0,0 +1,351 @@
+#include "tiledb/sm/filter/filter_storage.h"
+#include "tiledb/sm/misc/logger.h"
+#include "tiledb/sm/misc/parallel_functions.h"
+#include "tiledb/sm/misc/stats.h"
+#include "tiledb/sm/tile/tile.h"
+
+namespace tiledb {
+namespace sm {
+
+FilterPipeline::FilterPipeline() {
+  current_tile_ = nullptr;
+}
+
+FilterPipeline::FilterPipeline(const FilterPipeline& other) {
+  for (auto& filter : other.filters_) {
+    std::unique_ptr<Filter> copy(filter->clone());
+    add_filter(std::move(copy));
+  }
+  current_tile_ = other.current_tile_;
+}
+
+FilterPipeline& FilterPipeline::operator=(const FilterPipeline& other) {
+  // Call copy constructor
+  FilterPipeline copy(other);
+  // Swap with the temporary copy
+  swap(copy);
+  return *this;
+}
+
+Status FilterPipeline::add_filter(std::unique_ptr<Filter> filter) {
+  filter->set_pipeline(this);
+  filters_.push_back(std::move(filter));
+  return Status::Ok();
+}
+
+Status FilterPipeline::compute_tile_chunks(
+    Tile* tile, std::vector<std::pair<void*, uint32_t>>* chunks) const {
+  // For coordinate tiles, we treat each dimension separately (chunks won't
+  // cross dimension boundaries, since the coordinates have been split).
+  // Attribute tiles are treated as a whole.
+  auto dim_num = tile->stores_coords() ? tile->dim_num() : 1;
+  auto dim_tile_size = tile->size() / dim_num;
+  auto dim_cell_size = tile->cell_size() / dim_num;
+
+  // Compute a chunk size as a multiple of the cell size.
+  uint64_t chunk_size = std::min(constants::max_tile_chunk_size, dim_tile_size);
+  chunk_size = chunk_size / dim_cell_size * dim_cell_size;
+  if (chunk_size > std::numeric_limits<uint32_t>::max())
+    return LOG_STATUS(
+        Status::FilterError("Filter error; chunk size exceeds uint32_t"));
+
+  // Compute number of chunks
+  auto chunk_num = dim_tile_size / (chunk_size) +
+                   uint64_t(bool(dim_tile_size % (chunk_size)));
+
+  // Compute the chunks
+  tile->reset_offset();
+  for (uint64_t i = 0; i < dim_num; i++) {
+    for (uint64_t j = 0; j < chunk_num; j++) {
+      // Compute the actual size of the chunk (may be smaller at the end of the
+      // tile if the chunk size doesn't evenly divide).
+      auto size = static_cast<uint32_t>(
+          std::min(chunk_size, dim_tile_size - j * chunk_size));
+      chunks->emplace_back(tile->cur_data(), size);
+      tile->advance_offset(size);
+    }
+  }
+
+  return Status::Ok();
+}
+
+const Tile* FilterPipeline::current_tile() const {
+  return current_tile_;
+}
+
+Status FilterPipeline::filter_chunks_forward(
+    const std::vector<std::pair<void*, uint32_t>>& chunks,
+    Buffer* output) const {
+  // Vector storing the output of the final pipeline stage for each chunk.
+  std::vector<std::pair<FilterBuffer, FilterBuffer>> final_stage_io(
+      chunks.size());
+
+  // Run each chunk through the entire pipeline.
+  parallel_for(0, chunks.size(), [&](uint64_t i) {
+    // TODO(ttd): can we instead allocate one FilterStorage per thread?
```

Comment:
Noting here that probably the easiest way forward is to do static work partioning and allocate one Filter storage class per work partition (thread).  As the work across threads is equally proportional, it should perform well.